http-server-tck/src/main/java/io/micronaut/http/server/tck/tests/constraintshandler/ControllerConstraintHandlerTest.java

```diff
@@ -0,0 +1,273 @@
+            HttpRequest.POST("/constraints-via-handler/with-non-null", "{\"username\":\"\",\"password\":\"secret\"}"),
+            (server, request) -> AssertionUtils.assertThrows(server, request, constraintAssertion("must not be blank\"")));
+
+
+        asserts(SPEC_NAME,
+            HttpRequest.POST("/constraints-via-on-error-method/with-non-null", "{\"username\":\"\",\"password\":\"secret\"}"),
+            (server, request) -> AssertionUtils.assertThrows(server, request, TEAPOT_ASSERTION));
+
+        asserts(SPEC_NAME,
+            HttpRequest.POST("/constraints-via-on-error-method/with-non-null", "{\"password\":\"secret\"}"),
+            (server, request) -> AssertionUtils.assertThrows(server, request, TEAPOT_ASSERTION));
+    }
+
+
+    private static HttpResponseAssertion constraintAssertion(String expectedMessage) {
+        return HttpResponseAssertion.builder()
+            .status(HttpStatus.BAD_REQUEST)
+            .assertResponse(response -> {
+                Optional<String> json = response.getBody(Argument.of(String.class));
+                assertTrue(json.isPresent());
+                assertTrue(json.get().contains(expectedMessage));
+            }).build();
+    }
+
+    @Controller("/constraints-via-handler")
+    @Requires(property = "spec.name", value = SPEC_NAME)
+    static class BodyController {
+        @Post
+        @Produces(MediaType.TEXT_PLAIN)
+        @Status(HttpStatus.OK)
+        void login(@Body @NotNull @Valid CredentialsWithoutNullabilityAnnotation credentials) {
+
+        }
+
+        @Post("/with-at-nullable")
+        @Produces(MediaType.TEXT_PLAIN)
+        @Status(HttpStatus.OK)
+        void login(@Body @NotNull @Valid CredentialsWithNullable credentials) {
+
+        }
+
+        @Post("/with-non-null")
+        @Produces(MediaType.TEXT_PLAIN)
+        @Status(HttpStatus.OK)
+        void login(@Body @NotNull @Valid CredentialsWithNonNull credentials) {
+
+        }
+    }
+
+    @Controller("/constraints-via-on-error-method")
+    @Requires(property = "spec.name", value = SPEC_NAME)
+    static class OnErrorMethodController {
+        @Post
+        @Produces(MediaType.TEXT_PLAIN)
+        @Status(HttpStatus.OK)
+        void login(@Body @NotNull @Valid CredentialsWithoutNullabilityAnnotation credentials) {
+        }
+
+        @Post("/with-at-nullable")
+        @Produces(MediaType.TEXT_PLAIN)
+        @Status(HttpStatus.OK)
+        void loginWithNullable(@Body @NotNull @Valid CredentialsWithNullable credentials) {
+        }
+
+        @Post("/with-non-null")
+        @Produces(MediaType.TEXT_PLAIN)
+        @Status(HttpStatus.OK)
+        void loginWithNullable(@Body @NotNull @Valid CredentialsWithNonNull credentials) {
+        }
+
+        @Error(exception = ConstraintViolationException.class)
+        @Status(HttpStatus.I_AM_A_TEAPOT)
+        Optional<Map> constraintsEx(ConstraintViolationException e, HttpRequest<?> request) {
+            return request.getBody(Map.class);
```

Comment:
this is not allowed. you can't access the body as another type because it has already been parsed to the body parameter type, even though that failed due to constraint validations.

Reply:
Is this an intentedchange of behaviour? this was allowed in Micronaut Framework 4?

Reply:
If I changed this method to: 

```
   Optional<Map> constraintsEx(ConstraintViolationException e, HttpRequest<?> request) {
            Optional<?> objectOptional = request.getBody();
            if (objectOptional.isEmpty()) {
                return Optional.empty();
            }
            Object obj = objectOptional.get();
            String password = null;
            if (obj instanceof CredentialsWithoutNullabilityAnnotation credentials) {
                password = credentials.getPassword();
            } else if (obj instanceof CredentialsWithNullable credentials) {
                password = credentials.getPassword();
            } else if (obj instanceof CredentialsWithNonNull credentials) {
                password = credentials.getPassword();
            }
            return password != null ? Optional.of(Map.of("password", password)) : Optional.empty();
        }
```

I get an error as: 

`10:55:23.473 [default-nioEventLoopGroup-5-3] ERROR i.m.http.server.RouteExecutor - Unexpected error occurred: Already claimed`

@yawkat please advise. What is the intended way to do this?

Reply:
@sdelamo i looked at it, the problem is that once the body has been bound with `@Body`, the body ownership is transferred to that method. This is important because if the parameter type is ByteBuffer for example, the method is now responsible for releasing it. At that point it's no longer valid to call request.getBody.

I think we could still allow this access technically, though it would be unsafe and may have perf implications.

Reply:
I guess the question is why the previously parsed body is not available via `getBody()` and instead the method is trying to parse the body again?

Reply:
ok in that case I guess the issue that @sdelamo is trying to solve now is how to get hold of the body that was decoded but failed validation later

Reply:
yes, we need to give the users the ability to bind the body

Reply:
@sdelamo i applied your change and mine, the test passes now